### PR TITLE
Refactor doc-strings in particle.param

### PR DESCRIPTION
**Pull request by steindev**
Rephrase for consistency between manipulators and
expand some of the comments for clarification of
the purpose of a specific manipulator.

### CODE REVIEW

**Comment by steindev** [comment]
Well, it just came to my mind, that I should also update all `particle.param`s in the examples and benchmarks, right?

**Review by psychocoderHPC — CHANGES_REQUESTED** [review]
Please do not forget to update all particle.param files in the example

**Comment by PrometheusPi** [comment]
@steindev What is the status of updating the remaining `particle.param` files?

**Review by PrometheusPi — DISMISSED** [review]

**Comment by psychocoderHPC** [comment]
> @steindev What is the status of updating the remaining `particle.param` files?

I suggested him to wait until this file is reviewed and then update the other files to avoid searching and updating all files again if you have a request.

**Comment by steindev** [comment]
Yes @PrometheusPi, if you have comments, the time to raise these is now.

**Comment by PrometheusPi** [comment]
@steindev I am already happy with the current status and approved the current version.

**Comment by steindev** [comment]
@PrometheusPi @psychocoderHPC I updated the doc strings of `startposition::` functors as well. Please have a look whether you like it or not and provide feedback. After your feedback I will fix the formatting and update all `particle.param`s in examples and benchmarks, too.

**Comment by steindev** [comment]
@PrometheusPi please have a look at the additional changes I made to `particle.param`

**Comment by PrometheusPi** [comment]
@steindev please fix your c++ style - CI says "no"

**Comment by steindev** [comment]
> @steindev please fix your c++ style - CI says "no"

Well, if this is your only comment I will fix the formatting and then transfer the new files to all setups.

**Comment by PrometheusPi** [comment]
@steindev Sorry - this is (probably) not my only comment - I just wanted the CI to pass before I have a look because otherwise I have to review again.

File: include/picongpu/param/particle.param [hunk 1]
```diff
@@ -46,42 +46,51 @@ namespace picongpu
...
+         * macro-particles which sample the evolution of the particle distribution
+         * function in phase space.
          */
         constexpr uint32_t TYPICAL_PARTICLES_PER_CELL = 2u;
 
         namespace manipulators
         {
-            /** Parameter for DriftParam
+            /** Define initial particle drift direction vector.
+             *  Needs to be normalized to 1!
```
Comment (PrometheusPi) [inline include/picongpu/param/particle.param]: This comment is not needed. The direction is normalized automatically [here](https://github.com/ComputationalRadiationPhysics/picongpu/blob/f2cfee4166013760bb387e3cf04c4091eecb1fa9/include/picongpu/particles/manipulators/unary/Drift.hpp#L68).

File: include/picongpu/param/particle.param [hunk 2]
```diff
@@ -46,42 +46,51 @@ namespace picongpu
...
+             *  This vector is used in struct DriftParam.
              */
             CONST_VECTOR(float_X, 3, DriftParam_direction, 1.0, 0.0, 0.0);
-            /** Parameter for a particle drift assignment
-             */
+
+            /** Define initial particle drift. */
             struct DriftParam
             {
                 static constexpr float_64 gamma = 1.0;
```
Comment (PrometheusPi) [inline include/picongpu/param/particle.param:67]: This is not a drift - this is no motion. Perhaps we should use a meaningful value.

File: include/picongpu/param/particle.param [hunk 3]
```diff
@@ -46,42 +46,51 @@ namespace picongpu
...
             };
-            /* definition a temperature assignment manipulator */
+            /** Definition of manipulator assigning a temperature
+             *  using parameters from struct TemperatureParam.
+             */
             using AddTemperature = unary::Temperature<TemperatureParam>;
 
-            //! Functor for a temperature assignment, follows requirements of param::TemperatureFunctor
+
+            /** Define initial particle temperature in dependence of position.
```
Comment (PrometheusPi) [inline include/picongpu/param/particle.param]:
```suggestion
            /** Define initial particle temperature as a function of position.
```

File: include/picongpu/param/particle.param [hunk 4]
```diff
@@ -117,57 +128,82 @@ namespace picongpu
...
                 }
             };
-
-            /** definition of a free particle manipulator: double weighting */
+            /** Definition of the free particle manipulator which doubles each weighting.*/
             using DoubleWeighting = generic::Free<DoubleWeightingFunctor>;
 
+
+            /** Define mask which randomly marks macro-particles used
+             *  by Radiation plugin to calculate far field radiation.
```
Comment (PrometheusPi) [inline include/picongpu/param/particle.param]:
```suggestion
             *  by the radiation plugin to calculate far field radiation.
```

File: include/picongpu/param/particle.param [hunk 5]
```diff
@@ -117,57 +128,82 @@ namespace picongpu
...
                 static constexpr uint32_t numParticlesPerCell = TYPICAL_PARTICLES_PER_CELL;
             };
-            /** definition of random particle position */
+            /** Definition of start position functor that randomly distributes macro-particles within a cell. */
             using Random = RandomImpl<RandomParameter>;
 
-            /** definition of random particle position and weighting, same template parameter */
+
+            /** Definition of start position functor that randomly distributes macro-particles within a cell
+             * and randomly distributes weighting across macro-particles within a cell.
```
Comment (PrometheusPi) [inline include/picongpu/param/particle.param]: not sure - should it be weighting**s** ?

File: include/picongpu/param/particle.param [hunk 6]
```diff
@@ -117,57 +128,82 @@ namespace picongpu
...
+            */
             using RandomPositionAndWeighting = RandomPositionAndWeightingImpl<RandomParameter>;
 
+
+            /** Define target number for marco-particles per cell along a direction.
+             * To be used in Quiet start position functor.
+             *
+             * Here, one macro-particle per cell along x, one macro-particle per cell along z, and
+             * TYPICAL_PARTICLES_PER_CELL macro-particle per cell along y.
```
Comment (PrometheusPi) [inline include/picongpu/param/particle.param]:
```suggestion
             * TYPICAL_PARTICLES_PER_CELL macro-particles per cell along y.
```

**Review by PrometheusPi — CHANGES_REQUESTED** [review]
Only minor changes

File: include/picongpu/param/particle.param (hunk 2 above)
Comment (steindev) [reply include/picongpu/param/particle.param:67]: I specified what is defined. The rephrased sentence includes the possibility of zero drift.

File: include/picongpu/param/particle.param (hunk 3 above)
Comment (steindev) [reply include/picongpu/param/particle.param]: Done.

File: include/picongpu/param/particle.param (hunk 4 above)
Comment (steindev) [reply include/picongpu/param/particle.param]: Done

File: include/picongpu/param/particle.param (hunk 5 above)
Comment (steindev) [reply include/picongpu/param/particle.param]: Done

**Comment by steindev** [comment]
CI passed in less than 20min. Is this the new normal @psychocoderHPC?

**Comment by steindev** [comment]
@PrometheusPi @psychocoderHPC PR is ready for review. All changes that were initially requested are integrated.

File: share/picongpu/examples/atomicPhysics/include/picongpu/param/particle.param [hunk 7]
```diff
@@ -1,5 +1,5 @@
-/* Copyright 2013-2020 Axel Huebl, Rene Widera, Marco Garten, Benjamin Worpitz,
- *                     Richard Pausch
+/* Copyright 2013-2022 Axel Huebl, Rene Widera, Benjamin Worpitz,
+ *                     Richard Pausch, Klaus Steiniger
```
Comment (PrometheusPi) [inline share/picongpu/examples/atomicPhysics/include/picongpu/param/particle.param:2]: @steindev why did you remove Marco?

**Review by PrometheusPi — CHANGES_REQUESTED** [review]
Just one error

**Comment by PrometheusPi** [comment]
Furthermore, it is not just a change of comments/documentation anymore—you renamed variables - thus I will remove the ci:no-compile flag.

File: share/picongpu/examples/atomicPhysics/include/picongpu/param/particle.param (hunk 7 above)
Comment (PrometheusPi) [reply share/picongpu/examples/atomicPhysics/include/picongpu/param/particle.param:2]: ping @steindev

Comment (PrometheusPi) [reply share/picongpu/examples/atomicPhysics/include/picongpu/param/particle.param:2]: ping @steindev

**Comment by steindev** [comment]
@PrometheusPi Finally made the remaining fix :roll_eyes:

**Comment by steindev** [comment]
Ohh wow, seems like I made a lot of errors :-(

**Comment by PrometheusPi** [comment]
@steindev Your code consistently fails with a few CI jobs. I restarted them, but they failed again.

**Comment by steindev** [comment]
@PrometheusPi Ready to be merged!

File: share/picongpu/tests/CollisionsBeamRelaxation/include/picongpu/param/particle.param [hunk 8]
```diff
@@ -134,17 +161,29 @@ namespace picongpu
...
 
         } // namespace startPosition
+
+        /** Number of maximum macro-particles per cell during density profile evaluation.
+         *
+         * Determines the weighting of a macro particle as well as the number of
+         * macro-particles which sample the evolution of the particle distribution
+         * function in phase space.
+         */
+        constexpr uint32_t TYPICAL_PARTICLES_PER_CELL = 100u;
```
Comment (PrometheusPi) [inline share/picongpu/tests/CollisionsBeamRelaxation/include/picongpu/param/particle.param]: Why is this defined below the definition in line 175 and 176 and thus does not reuse number?

File: share/picongpu/tests/CollisionsBeamRelaxation/include/picongpu/param/particle.param [hunk 9]
```diff
@@ -91,38 +100,56 @@ namespace picongpu
...
 
-            struct IonCharge3Impl
+            /** Define initial ion charge to thrice ionized*/
+            struct ThriceIonizedImpl
             {
+                /** Ionize ions thrice by removing three bound electrons */
                 template<typename T_Particle>
                 DINLINE void operator()(T_Particle& particle)
                 {
-                    particle[boundElectrons_] = 3.0_X;
```
Comment (PrometheusPi) [inline share/picongpu/tests/CollisionsBeamRelaxation/include/picongpu/param/particle.param:111]: How did this work before?

Comment (PrometheusPi) [reply share/picongpu/tests/CollisionsBeamRelaxation/include/picongpu/param/particle.param:111]: Okay - it's Carbon 6 - 3 = 3

File: share/picongpu/tests/CollisionsThermalisation/include/picongpu/param/particle.param [hunk 10]
```diff
@@ -115,18 +133,29 @@ namespace picongpu
...
 
         } // namespace startPosition
 
+        /** Number of maximum macro-particles per cell during density profile evaluation.
+         *
+         * Determines the weighting of a macro particle as well as the number of
+         * macro-particles which sample the evolution of the particle distribution
+         * function in phase space.
+         */
+        constexpr uint32_t TYPICAL_PARTICLES_PER_CELL = 1000u;
```
Comment (PrometheusPi) [inline share/picongpu/tests/CollisionsThermalisation/include/picongpu/param/particle.param]: same here

**Review by PrometheusPi — CHANGES_REQUESTED** [review]
Only a minor change requested: 
Typical PPC number should either be determined automatically or used as input. Currently, it is defined multiple times.

File: include/picongpu/param/particle.param [hunk 11]
```diff
@@ -46,42 +46,50 @@ namespace picongpu
...
-        /** Number of maximum particles per cell during density profile evaluation.
+        /** Number of maximum macro-particles per cell during density profile evaluation.
          *
-         * Determines the weighting of a macro particle and with it, the number of
-         * particles "sampling" dynamics in phase space.
+         * Determines the weighting of a macro particle as well as the number of
+         * macro-particles which sample the evolution of the particle distribution
+         * function in phase space.
          */
         constexpr uint32_t TYPICAL_PARTICLES_PER_CELL = 2u;
```
Comment (BrianMarre) [inline include/picongpu/param/particle.param]: TYPICAL_NUMBER_PARTICLE_PER_CELL [does not influence weighting of a particles](https://github.com/ComputationalRadiationPhysics/picongpu/blob/f1f9a2d5781a53469a2124dbaa5872a9e8aaba49/include/picongpu/particles/InitFunctors.hpp#L58-L68).
It is only used in the normalization of our internal unit system, but we do not normalize macro particle weighting to some reference value currently (we instead use the number of physical particles directly).

File: share/picongpu/tests/CollisionsBeamRelaxation/include/picongpu/param/particle.param (hunk 9 above)
Comment (BrianMarre) [reply share/picongpu/tests/CollisionsBeamRelaxation/include/picongpu/param/particle.param:111]: just as a discussion point, we have a specific [functor](https://github.com/ComputationalRadiationPhysics/picongpu/blob/f1f9a2d5781a53469a2124dbaa5872a9e8aaba49/include/picongpu/particles/atomicPhysics/SetToAtomicGroundStateForChargeState.hpp#L24-L33) for that initialization.
Used as follows:
```
namespace picongpu::particles::manipulators
{
    struct SetIonTo16Ionized
    {
        /** set ions to targetChargeState times ionized and atomic ground state for
         *  that charge state.
         */
        template<typename T_Particle>
        DINLINE void operator()(T_Particle& particle)
        {
            constexpr float_X targetChargeState = 16;

            // number of Electrons for neutral atom
            constexpr float_X numberBoundElectrons
                = picongpu::traits::GetAtomicNumbers<T_Particle>::type::numberOfProtons;

            PMACC_ASSERT_MSG(
                numberBoundElectrons >= targetChargeState,
                "too few protons for target charge state");

            picongpu::particles::atomicPhysics::SetToAtomicGroundStateForChargeState{}(
                particle,
                numberBoundElectrons - targetChargeState);
        }
    };

    using SetIonization = generic::Free<SetIonTo16Ionized>;

} // namespace picongpu::particles::manipulators
```
(works in principle the same as the old solution if we only track charge states, but ensures consistency between charge state and atomic state if present). We(I) already switched all ionization to the new interface, but did not switch the particle.param files yet.
Do we wait for with this change or do we want to do it now?

File: share/picongpu/tests/CollisionsThermalisation/include/picongpu/param/particle.param (hunk 10 above)
Comment (BrianMarre) [reply share/picongpu/tests/CollisionsThermalisation/include/picongpu/param/particle.param]: this line moved from line 54, but is also seems to have changed, original value was 350 now it is 1000, @steindev was there any particular reason for that?

File: share/picongpu/tests/CollisionsBeamRelaxation/include/picongpu/param/particle.param (hunk 8 above)
Comment (steindev) [reply share/picongpu/tests/CollisionsBeamRelaxation/include/picongpu/param/particle.param]: Either of the two definitions in 175 & 176 can be chosen in the simulation. How should I know at this point which of these is used? So I chose the smaller value.

File: share/picongpu/tests/CollisionsBeamRelaxation/include/picongpu/param/particle.param (hunk 9 above)
Comment (steindev) [reply share/picongpu/tests/CollisionsBeamRelaxation/include/picongpu/param/particle.param:111]: Included assert and usage of functor as Brian suggests

File: share/picongpu/tests/CollisionsBeamRelaxation/include/picongpu/param/particle.param (hunk 8 above)
Comment (steindev) [reply share/picongpu/tests/CollisionsBeamRelaxation/include/picongpu/param/particle.param]: I guess, in principle I could know from the preprocessor variable `PARAM_START_ELECTRONS` defined in `speciesInitialization.param`, but I do not know which file is included first. I could look this up, but I would add quite a number of comments to these two files in order to ensure consistency. And all this for a single variable which is not really important, i.e. its exact value does not matter.

File: share/picongpu/tests/CollisionsThermalisation/include/picongpu/param/particle.param (hunk 10 above)
Comment (steindev) [reply share/picongpu/tests/CollisionsThermalisation/include/picongpu/param/particle.param]: Same as above. I probably changed it in order to match it to at least one of the possible values.

File: include/picongpu/param/particle.param (hunk 11 above)
Comment (steindev) [reply include/picongpu/param/particle.param]: Ahhh, :cursing_face:. Thanks.

File: share/picongpu/tests/CollisionsBeamRelaxation/include/picongpu/param/particle.param (hunk 8 above)
Comment (PrometheusPi) [inline share/picongpu/tests/CollisionsBeamRelaxation/include/picongpu/param/particle.param]: How about? 

```suggestion
        /** Number of maximum macro-particles per cell during density profile evaluation.
         *
         * Determines the weighting of a macro particle as well as the number of
         * macro-particles which sample the evolution of the particle distribution
         * function in phase space.
         */
        constexpr uint32_t TYPICAL_PARTICLES_PER_CELL = 100u;

            /** Definition of start position functor that
             * places macro-particles at the initial in-cell position defined above.
             * Number of macro-particles per cell is defined by argument to OnePositionParam<>.
             */
            using Corner = OnePositionImpl<OnePositionParam<TYPICAL_PARTICLES_PER_CELL>>;

        } // namespace startPosition

```

File: include/picongpu/param/particle.param (hunk 11 above)
Comment (BrianMarre) [reply include/picongpu/param/particle.param]: don't worry, I built two separate bugs into FLYonPIC because I did not know either

File: share/picongpu/tests/CollisionsBeamRelaxation/include/picongpu/param/particle.param (hunk 8 above)
Comment (steindev) [reply share/picongpu/tests/CollisionsBeamRelaxation/include/picongpu/param/particle.param]: This does not work, both `Corner100ppc` and `Corner1000ppc` are used in the simulation. See `speciesInitialization.param` (https://github.com/steindev/picongpu/blob/390789d086a4c96c36f0b24d9da2af9516feeac8/share/picongpu/tests/CollisionsBeamRelaxation/include/picongpu/param/speciesInitialization.param#L50).

File: include/picongpu/param/particle.param (hunk 11 above)
Comment (steindev) [reply include/picongpu/param/particle.param]: I moved this documentation text to the numParticlesPerCell variable within the startPosition functors.

File: share/picongpu/tests/CollisionsBeamRelaxation/include/picongpu/param/particle.param (hunk 8 above)
Comment (steindev) [reply share/picongpu/tests/CollisionsBeamRelaxation/include/picongpu/param/particle.param]: Done. (According to offline discussion)

**Review by PrometheusPi — APPROVED** [review]

**Comment by PrometheusPi** [comment]
@psychocoderHPC are you still against merging this pull request?

**Review by psychocoderHPC — APPROVED** [review]
